### ISSUE TITLE
Event MessageAdded usually does not be fired!

### ISSUE THREAD

**Issue by tranty9597** [twilio/TwilioChatJsReactNative#31]
Acture behavior:

Event "MessageAdded" ussally does not be fired. It casuse the issue that each member see the difference messages.

Step to procedure:

1. Create channel
2. Listen to event messageAdded on this channel
3. Send serires message

Version
React Native 0.60.4
Twilio-Chat 3.3.1

**Comment by aleksandrsivanovs** [comment]
Hi!

we haven't observed such behaviour. 
Can you provide any more debugging information regarding this?

Aleksandrs

**Comment by tranty9597** [comment]
> Hi!
> 
> we haven't observed such behaviour.
> Can you provide any more debugging information regarding this?
> 
> Aleksandrs

I put console.log at sendMessag, client.on(MessageAdded), channel.onMessageAdded to log message index that has been sent.

Hear is debug console

![Screen Shot 2019-08-09 at 8 55 37 AM](https://user-images.githubusercontent.com/41298303/62748711-d47db100-ba83-11e9-8f44-6a830a241c6c.png)

![Screen Shot 2019-08-09 at 9 09 55 AM](https://user-images.githubusercontent.com/41298303/62749138-7fdb3580-ba85-11e9-8a1b-eb67537e535a.png)

You can see the message Index 65 and 95 has been sent but messageAdded does not be fired at client and channel either.

**Comment by spalmertwilio** [comment]
@aleksandrsivanovs please see IPM-2021 - I think the user for that issue is hitting a race condition due to their steps.  But the create messages very quickly to display the issue this user mentions here.

**Comment by tranty9597** [comment]
> @aleksandrsivanovs please see IPM-2021 - I think the user for that issue is hitting a race condition due to their steps. But the create messages very quickly to display the issue this user mentions here.

When i use Twilio-Chat for my React Js, this issue is not happened. 

In my React Native project: I have a TwiChannelWrapper class to handle Channel Event and forward it to view. My view listens to TwiChannelWrapper to update view. 

In my React JS project: I'm using direcly Channel of Twilio Chat to listen event for view updating.

**Comment by tranty9597** [comment]
Do we have any difference bettween Test Account and Premium Account?

**Comment by aleksandrsivanovs** [comment]
there should not be any difference between test and premium accounts.

the difference of wrapper and using the channel class directly is interesting. 
Is it possible that callback on wrapper is not executing because it is blocked / processing previous callback  (I do not have such deep knowledge of RN)?

**Comment by tranty9597** [comment]
> there should not be any difference between test and premium accounts.
> 
> the difference of wrapper and using the channel class directly is interesting.
> Is it possible that callback on wrapper is not executing because it is blocked / processing previous callback (I do not have such deep knowledge of RN)?

I've put console log at event listener of twilio channel and client not my wrapper. 
![Screen Shot 2019-08-23 at 2 40 25 PM](https://user-images.githubusercontent.com/41298303/63575344-0cbcdd80-c5b4-11e9-93cb-b7a2ac8b7996.png)
![Screen Shot 2019-08-23 at 2 43 01 PM](https://user-images.githubusercontent.com/41298303/63575460-54436980-c5b4-11e9-836e-ab683b60b0ff.png)

**Comment by aleksandrsivanovs** [comment]
and it still misses messageAdded in case of rapid message sending?
(one more question: are you sending message from the same client or from different client?)

**Comment by tranty9597** [comment]
Yep, it still misses. I send message with only one client in one user session.

**Comment by tranty9597** [comment]
More info: I have config for push notification for each added  message and notification is not missed.

**Comment by aleksandrsivanovs** [comment]
I've created branch with test which tests purely twilio-chat: https://github.com/twilio/TwilioChatJsReactNative/tree/rapid-message-send-test
(the test itself is https://github.com/twilio/TwilioChatJsReactNative/blob/rapid-message-send-test/__tests__/rapidMessageSend.js)

I can not reproduce this problem through tests.

Can it be that your app somehow locks somewhere and in the result it "loses" the event on JS side?
The reason I'm asking is that as you've mentioned, pushes are arriving fine on device, which means that backend actually sends all the notifications in time and push rendering part (managed by OS) is working fine.

**Comment by aleksandrsivanovs** [comment]
^ merged to master now: https://github.com/twilio/TwilioChatJsReactNative/blob/master/__tests__/messages.js

**Comment by tootsweet** [comment]
Hello,


I dug into the Twilio JS code and I believe I have identified this bug, or at least one scenario that allows it to happen. Here’s how it goes (the React Native app is Bob, Alice is another client, in my case, a browser).

Please note that the JS chat SDK uses an implementation of a SyncList to deal with the messages of a given channel (get, add, remove, etc).

• Alice sends “1”
• Bob’s synclist receives “1” and notifies the app
• Alice sends “2”
• Bob’s synclist receives “2” and notifies the app
• Alice sends “3”
• Bob’s synclist receives “3” and notifies the app
• Alice sends “4”
• Bob calls the channel.getMessages(1) function in order to get the last consumed message, triggering the SyncList.queryItems function which puts message “4” into the list’s cache
• Subsequently, the chat SDK does not notify the app because the “4” message is already cached, so it assumes that it already has been
• Alice sends “5”
• Bob’s synclist receives “5” and notifies the app

Why am I calling channel.getMessages(1)? It’s because the "unconsumed messages” feature is too slow and unreliable, so I had to replace it by the following mechanism:

• Fetch the last message of the channel
• Compare its index to the channel.lastConsumedIndex field
• The difference between the two indices is the number of unconsumed messages

The fact that a simple read operation (getMessages) can change the SDK’s state is quite surprising, and this should not happen.

What happens if somebody query the messages of a channel while other people are sending messages?

If the getMessages call finishes before the SDK notifies its listeners of the new messages, then no notifications will be sent for those new messages.

This is a major bug.

Can you guys fix this?

Regards,
Elyes GHERIB
CTO @ Toot Sweet

**Comment by aleksandrsivanovs** [comment]
Hi!

First of all - great analysis, we were impressed of such investigation details.
 
Let's divide problem you are having into two areas.
1) unconsumed messages count.
Why do you say that it's slow and unreliable?
It should not be slower than the call to get messages.
In general, due to the internals currently it is the only way to get most reliable unread messages count. 
Your model is not working for counting unconsumed messages because of the flaw in your algorithm: by the nature of our backend the messages index is guaranteed to be increasing with each new message, however, this is not monotonously increasing. If you would take a look into Twilio Sync's documentation, you can notice that for Channel's Messages we are using the Sync's List concept. It states that every event has index, i.e. add, update, delete. So, we can't guarantee monotonous increase for message index, we can guarantee only increase. 
In general, this breaks your algorithm to calculate unread messages count: Channel.getMessages(1).index - Channel. lastConsumedMessageIndex. That will yield number which is equal or greater than actual unconsumed messages count.
We advice to use Channel.getUnconsumedMessagesCount() and figure out why it won't work for you. 

2) Mentioned Sync's "bug" in getMessages() vs messageAdded events.
This is not actually a bug. 
Our promise is to deliver Channel's Messages to the client. That can be done with push (via events) or pull (via getMessages()) model. 
In your case if you are pulling messages (with Channel.getMessages(1)) you will get message as the result of that call. And, anything which you won't pull - we will deliver via messageAdded event. 
If we were to enable messageAdded event for all incoming messages, without looking as Channel.getMessages() result - it would result in problem of solving deduplication of messages (one duplicate coming from getMessages() and second arrived through "messageAdded") which then is completely in the hands of developer, which results in unclear and hard to use SDK. If you insist on your usecase - then you can simply pass the result of Channel.getMessages() to the same handler you have for messageAdded event. 
Our assumption is that most people won't have to implement more complex logic on the client side, it will "just work" for them. 

So, in short - let's work out why Channel.getUnconsumedMessagesCount() is not working for you :)

Aleksandrs

**Comment by tootsweet** [comment]
Hello Aleksandrs,


Thanks! 

Here's my use case:
- sender is a browser (Alice)
- receiver is a React Native app (Bob)
- Alice sends 5 messages very quickly
- Bob receives all 5 messages
- Bob calls getUnconsumedMessagesCount(), gets 1 or 2, whereas the correct value is 5
- Bob receives 5 "channelUpdated" callbacks, all with the same update reason: "lastMessage".

If I add an extra step:
- Bob waits 3 seconds then calls getUnconsumedMessagesCount()

Then I get a correct result.

There's my problem with getUnconsumedCount(): it is not updated reliably when messages come in fast.

I can show you a live demo if you want.


Elyes

**Comment by aleksandrsivanovs** [comment]
Ah, for that we know the cause: in the SDK there is cache with 1 second lifetime.
I see that you've opened twilio support ticket - let's continue this discussion there, as the unconsumed messages count is not related to the event messageAdded. 

I will close this item here on github.com then.

**Comment by bilalahmed13** [comment]
hi @aleksandrsivanovs 
i too have a similar problem. the "messageadded" event was working quite well and suddenly it stopped working.
i can saw the new messages in the twilio dashboard messages for the channel but since the "messageadded" event stopped working for the client, i cannot red these messages in the client side. i am using node js for twilio client.
i have to re run my script and it started working again. any help in this regard is much appreciated.
